core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -99,20 +140,54 @@ public void registerCacheEventListener(CacheEventListener<? super K, ? super V>
    */
   @Override
   public void deregisterCacheEventListener(CacheEventListener<? super K, ? super V> listener) {
-    if (!registeredListeners.remove(new EventListenerWrapper(listener,
-        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class)))) {
-      throw new IllegalStateException("Unknown cache event listener: " + listener);
+    EventListenerWrapper wrapper = new EventListenerWrapper(listener,
+        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class));
+    boolean removed = false;
+    for (EventListenerWrapper listenerWrapper : aSyncListenersSet) {
+      if(listenerWrapper.equals(wrapper)) {
+        aSyncListenersSet.remove(listenerWrapper);
+        removed = true;
+        if(listenerWrapper.config.orderingMode() == EventOrdering.ORDERED) {
+          orderedListenerCount.decrementAndGet();
+        }
+        break;
+      }
     }
-    if (!hasListeners()) {
-      store.disableStoreEventNotifications();
+    if (!removed) {
+      for (EventListenerWrapper listenerWrapper : syncListenersSet) {
+        if(listenerWrapper.equals(wrapper)) {
+          syncListenersSet.remove(listenerWrapper);
+          removed = true;
+          if(listenerWrapper.config.orderingMode() == EventOrdering.ORDERED) {
+            orderedListenerCount.decrementAndGet();
+          }
+          break;
+        }
+      }
+    }
+    synchronized (this) {
+      if (!hasListeners()) {
+        if (eventThreadLocal instanceof EventThreadLocalImpl) {
+          eventThreadLocal = new NoOpEventThreadLocalImpl();
+        }
+        store.disableStoreEventNotifications();
+      }
+    }
+    if(!removed) {
+      throw new IllegalStateException("Unknown cache event listener: " + listener);
     }
   }
 
   // TODO this should be really the shutdown method for the service
   @Override
   public void releaseAllListeners() {
```

Comment:
Objective – clear everything maintained about the listeners

What it internally does:
•	Reset orderedListeners count
•	Clear sync/async sets
•	Disable EventThreadLocal to a noop
Why it should work:
A reassignment of the threadLocal to a noop implementation ensures that an unnecessary threadLocal does not exist when there are no listeners are available.